src/Compile.purs

```diff
@@ -0,0 +1,346 @@
+      SolcError { sourceLocation
+                , type: _type
+                , severity
+                , message
+                , formattedMessage
+                }
+--------------------------------------------------------------------------------
+
+-- | This is the artifact we want, compatible with truffle (subset)
+newtype OutputContract =
+  OutputContract { abi :: A.JArray
+                 , bytecode :: String
+                 }
+
+parseOutputContract
+  :: A.Json
+  -> Either String OutputContract
+parseOutputContract json = do
+  obj <- A.decodeJson json
+  abi <- obj A..? "abi"
+  evm <- obj A..? "evm"
+  evmObj <- A.decodeJson evm
+  bytecodeO <- evmObj A..? "bytecode"
+  bytecode <- bytecodeO A..? "object"
+  pure $ OutputContract {abi, bytecode}
+
+instance encodeOutputContact :: A.EncodeJson OutputContract where
+  encodeJson (OutputContract {abi, bytecode}) =
+    "abi" A.:= A.fromArray abi A.~>
+    "bytecode" A.:= bytecode A.~>
+    A.jsonEmptyObject
+
+decodeContract
+  :: String
+  -> A.Json
+  -> Either String (M.StrMap OutputContract)
+decodeContract srcName json = do
+  obj <- A.decodeJson json
+  contracts <- obj A..? "contracts"
+  json' <- maybe (Left $ "couldn't find " <> show srcName <> " in object") Right (M.lookup srcName contracts)
+  contractMap <- A.decodeJson json'
+  for contractMap $ parseOutputContract
+
+foreign import jsonStringifyWithSpaces :: Int -> A.Json -> String
+
+writeBuildArtifact
+  :: forall eff m.
+     MonadAff (fs :: FS.FS | eff) m
+  => String
+  -> FilePath
+  -> A.Json
+  -> m Unit
+writeBuildArtifact srcName filepath output = liftAff $
+  case decodeContract srcName output of
+    Left err -> throwError <<< error $ "Malformed solc output: " <> err
+    Right co -> do
+      let dn = Path.dirname filepath
+      dnExists <- FS.exists dn
+      if not dnExists
+        then liftEff (mkdirp dn)
+        else do
+          isDir <- Stats.isDirectory <$> FS.stat dn
+          if not isDir
+            then (throwError <<< error $ ("Path " <> show dn <> " exists but is not a directory!"))
+            else pure unit
+      let parsedPath = Path.parse filepath
+          withNewExtension = Path.concat [parsedPath.dir, parsedPath.name <> ".json"]
+          contractsMainModule = M.lookup parsedPath.name co -- | HACK HACK HACK HACK HACK HACK HACK HACK HACK HACK
+      case contractsMainModule of -- | HACK HACK HACK HACK HACK HACK HACK HACK HACK HACK
+        Nothing -> (throwError <<< error $ ("Couldn't find an object named " <> show parsedPath.name <> " in " <> show filepath <> "!"))
+        Just co' -> FS.writeTextFile UTF8 withNewExtension <<< jsonStringifyWithSpaces 4 $ A.encodeJson co'
+
+
+--------------------------------------------------------------------------------
+
+newtype SolcOutput =
+  SolcOutput { errors :: Maybe (Array SolcError)
+             , contracts :: M.StrMap OutputContract -- mapping of Filepath
+             }
+
+{-
+
+
+{ "contractName": "SimpleStorage"
```

Comment:
can we remove?

Reply:
I meant the comment

Reply:
oh that’s the big comment blob. showed up as the HACKHACKHACK line for me :/. keep it in, will help with making the types truffle compatible for the aforementioned